File: apps/desktop/src/renderer/services/documents.ts
import { api } from "@/services/api";
import { scheduleAccountRulesMemoryRefresh } from "@/services/refreshAccountRulesMemory";

/**
 * Document tree REST client (`/v1/documents`) — unified md **entries** under the
 * convention root (Agent记忆与知识系统 · 目标形态「统一 md 条目基座」).
 *
 * Nodes are addressed by **id**. Scope is `folderId` (`null` = GLOBAL). The file rail
 * lists entries flat by scope (no 记忆/规则/文档 folders). `apply_mode` / `description`
 * are derived indexes of body frontmatter; UI shows 常驻/按需 badges, never the raw
 * `apply` key. `frontmatter_error` means the entry does not inject — surface it.
 *
 * Always-pool meter: {@link getAlwaysQuota}. Write past the cap while editing an
 * existing always entry returns `quota_warning`; create / promote past the cap is
 * 409 `ALWAYS_QUOTA_EXCEEDED`.
 *
 * 纠错通道: {@link setDocumentDisputed} marks an entry「这条不对」so the AI stops using
 * it while the entry itself stays readable (`disputedAt`).
 */

/** Cloud-documents convention root name (§5.0). ≠ local disk `~/Documents/AgentCore`. */
export const AGENTCORE_ROOT_NAME = "AgentCore";

/** Legacy rules directory under the convention root (server still parents new rules here). */
export const RULES_DIR_NAME = "规则";

/** Legacy memory directory under the convention root (AI-maintained notes). */
export const MEMORY_DIR_NAME = "记忆";

/**
 * User-facing injection mode (§5.4 / 目标形态). API may still store other values;
 * the desktop surface only offers these two.
 */
export type DocumentApplyMode = "always" | "on_demand";

/** Map wire `apply_mode` onto the two-state UI (unknown → on_demand per frontmatter default). */
export function toApplyMode(raw: string): DocumentApplyMode {
  return raw === "always" ? "always" : "on_demand";
}

/** A tree node's metadata (list rows — body omitted so a listing stays light). */
export interface DocumentNode {
  id: string;
  parentId: string | null;
  /** Scope: null = GLOBAL layer, else the project (folder) this entry is bound to. */
  folderId: string | null;
  kind: "folder" | "document";
  role: "rule" | "general";
  /** true = AI-maintained (write-side / UI review); false = user-owned. */
  aiMaintained: boolean;
  applyMode: DocumentApplyMode;
  /** One-line summary from frontmatter; empty is fine (not an error). */
  description: string;
  name: string;
  /** Structural frontmatter failure — entry does not inject; UI must report it. */
  frontmatterError: string | null;
  /**
   * Chars this entry contributes to the always pool (`null` when not always).
   * Matches server `always_entry_chars` so row totals == meter `usedChars`.
   */
  alwaysChars: number | null;
  /**
   * When the user marked this entry wrong (`null` = live). Disputed entries are kept and
   * still editable here, but the AI stops injecting / consulting them (纠错通道).
   */
  disputedAt: string | null;
}

/** A node plus its markdown body + content-hash CAS tag (the editor's load payload). */
export interface DocumentDetail extends DocumentNode {
  content: string;
  version: string;
  /** Soft warning when a user edit of an existing always entry exceeds the pool. */
  quotaWarning: string | null;
}

export interface DocumentWriteResult {
  ok: boolean;
  /** Content-addressed CAS tag; sent back as the next write's baseline (stale → conflict). */
  version: string;
  conflict: boolean;
  frontmatterError: string | null;
  quotaWarning: string | null;
}

/**
 * Always-pool usage for the UI meter.
 * `usedChars == globalChars + projectChars` (project scope includes global).
 */
export interface AlwaysQuota {
  usedChars: number;
  maxChars: number;
  percent: number;
  /** Global-scope always chars in this meter context. */
  globalChars: number;
  /** Project-scope always chars (0 when the meter is global-only). */
  projectChars: number;
}

interface DocumentNodeWire {
  id: string;
  parent_id: string | null;
  folder_id: string | null;
  kind: string;
  role: string;
  ai_maintained: boolean;
  apply_mode: string;
  description?: string | null;
  name: string;
  frontmatter_error?: string | null;
  always_chars?: number | null;
  disputed_at?: string | null;
}

interface DocumentDetailWire extends DocumentNodeWire {
  content: string;
  version: string;
  quota_warning?: string | null;
}

interface DocumentWriteWire {
  ok: boolean;
  version: string;
  conflict?: boolean;
  frontmatter_error?: string | null;
  quota_warning?: string | null;
}

interface AlwaysQuotaWire {
  used_chars: number;
  max_chars: number;
  percent: number;
  global_chars?: number;
  project_chars?: number;
}

const toNode = (w: DocumentNodeWire): DocumentNode => ({
  id: w.id,
  parentId: w.parent_id,
  folderId: w.folder_id,
  kind: w.kind === "folder" ? "folder" : "document",
  role: w.role === "rule" ? "rule" : "general",
  aiMaintained: w.ai_maintained,
  applyMode: toApplyMode(w.apply_mode),
  description: (w.description ?? "").trim(),
  name: w.name,
  frontmatterError: w.frontmatter_error?.trim() || null,
  alwaysChars:
    typeof w.always_chars === "number" && Number.isFinite(w.always_chars)
      ? w.always_chars
      : null,
  disputedAt: w.disputed_at?.trim() || null,
});

const toDetail = (w: DocumentDetailWire): DocumentDetail => ({
  ...toNode(w),
  content: w.content,
  version: w.version,
  quotaWarning: w.quota_warning?.trim() || null,
});

const toWriteResult = (w: DocumentWriteWire): DocumentWriteResult => ({
  ok: w.ok,
  version: w.version,
  conflict: Boolean(w.conflict),
  frontmatterError: w.frontmatter_error?.trim() || null,
  quotaWarning: w.quota_warning?.trim() || null,
});

/** List a folder's direct children (`parentId` null = the user's top-level nodes). */
export function listDocuments(
  parentId: string | null = null,
): Promise<DocumentNode[]> {
  const q = parentId ? `?parent_id=${encodeURIComponent(parentId)}` : "";
  return api
    .get<DocumentNodeWire[]>(`/v1/documents${q}`)
    .then((rows) => rows.map(toNode));
}

/**
 * Flat list of inject-able **entries** for one scope (global when `folderId` is null).
 * Walks `AgentCore/{规则,记忆}/` (and one nested level) plus leftover top-level docs.
 * Does **not** group by `role` — UI partitions by scope only.
 */
export async function listScopeEntries(
  folderId: string | null = null,
): Promise<DocumentNode[]> {
  const tops = await listDocuments(null);
  const byId = new Map<string, DocumentNode>();

  const takeDoc = (n: DocumentNode) => {
    if (n.kind === "document" && n.folderId === folderId) byId.set(n.id, n);
  };

  for (const n of tops) takeDoc(n);

  const agentcores = tops.filter(
    (n) =>
      n.kind === "folder" &&
      n.name === AGENTCORE_ROOT_NAME &&
      n.folderId === folderId,
  );

  await Promise.all(
    agentcores.map(async (ac) => {
      const kids = await listDocuments(ac.id);
      await Promise.all(
        kids.map(async (kid) => {
          if (kid.kind === "document") {
            takeDoc(kid);
            return;
          }
          if (
            kid.kind !== "folder" ||
            (kid.name !== RULES_DIR_NAME && kid.name !== MEMORY_DIR_NAME)
          ) {
            return;
          }
          const leaves = await listDocuments(kid.id);
          await Promise.all(
            leaves.map(async (leaf) => {
              if (leaf.kind === "document") {
                takeDoc(leaf);
                return;
              }
              if (leaf.kind !== "folder") return;
              const nested = await listDocuments(leaf.id);
              for (const n of nested) takeDoc(n);
            }),
          );
        }),
      );
    }),
  );

  return [...byId.values()].sort((a, b) => a.name.localeCompare(b.name, "zh"));
}

/**
 * All user-owned rule documents across scopes (legacy helper). Prefer
 * {@link listScopeEntries} for the flat file-rail list.
 */
export async function listUserRules(): Promise<DocumentNode[]> {
  const tops = await listDocuments(null);
  const byId = new Map<string, DocumentNode>();
  const isUserRule = (n: DocumentNode) =>
    n.role === "rule" && !n.aiMaintained && n.kind === "document";

  for (const n of tops) {
    if (isUserRule(n)) byId.set(n.id, n);
  }

  const agentcores = tops.filter(
    (n) => n.kind === "folder" && n.name === AGENTCORE_ROOT_NAME,
  );
  await Promise.all(
    agentcores.map(async (ac) => {
      const kids = await listDocuments(ac.id);
      const rulesDir = kids.find(
        (k) => k.kind === "folder" && k.name === RULES_DIR_NAME,
      );
      if (!rulesDir) return;
      const rules = await listDocuments(rulesDir.id);
      for (const n of rules) {
        if (isUserRule(n)) byId.set(n.id, n);
      }
    }),
  );

  return [...byId.values()].sort((a, b) => a.name.localeCompare(b.name, "zh"));
}

/** Always-pool usage for the injection context (global + optional project). */
export function getAlwaysQuota(
  folderId: string | null = null,
): Promise<AlwaysQuota> {
  const q =
    folderId != null ? `?folder_id=${encodeURIComponent(folderId)}` : "";
  return api
    .get<AlwaysQuotaWire>(`/v1/documents/always-quota${q}`)
    .then((w) => {
      const globalChars =
        typeof w.global_chars === "number" && Number.isFinite(w.global_chars)
          ? w.global_chars
          : folderId == null
            ? w.used_chars
            : 0;
      const projectChars =
        typeof w.project_chars === "number" && Number.isFinite(w.project_chars)
          ? w.project_chars
          : folderId != null
            ? Math.max(0, w.used_chars - globalChars)
            : 0;
      return {
        usedChars: w.used_chars,
        maxChars: w.max_chars,
        percent: w.percent,
        globalChars,
        projectChars,
      };
    });
}

/** Load one document's body + CAS version (the editor's load). */
export function getDocument(id: string): Promise<DocumentDetail> {
  return api
    .get<DocumentDetailWire>(`/v1/documents/${encodeURIComponent(id)}`)
    .then(toDetail);
}

/**
 * Create a user-owned entry in a scope (`folderId` null = global).
 * Server still parents under `AgentCore/规则/` until role 三分 is removed.
 */
export function createRuleDocument(
  name: string,
  folderId: string | null = null,
  content = "",
): Promise<DocumentDetail> {
  return api
    .post<DocumentDetailWire>("/v1/documents", {
      name,
      kind: "document",
      role: "rule",
      content,
      parent_id: null,
      folder_id: folderId,
      apply_mode: "always",
    })
    .then(toDetail)
    .then((doc) => {
      scheduleAccountRulesMemoryRefresh();
      return doc;
    });
}

/** Switch an entry's injection mode (`always` ↔ `on_demand`). */
export function updateDocumentApplyMode(
  id: string,
  applyMode: DocumentApplyMode,
): Promise<DocumentNode> {
  return api
    .patch<DocumentNodeWire>(`/v1/documents/${encodeURIComponent(id)}`, {
      apply_mode: applyMode,
    })
    .then(toNode)
    .then((node) => {
      scheduleAccountRulesMemoryRefresh();
      return node;
    });
}

/**
 * Mark an entry as wrong / undo that mark (纠错通道「这条不对」).
 *
 * Explicit user action only — nothing here is inferred from what was said in a
 * conversation. A disputed entry stops being injected and stops appearing in the AI's
 * on-demand catalog, but is neither deleted nor rewritten, so the user can read what was
 * wrong and undo the mark.
 */
export function setDocumentDisputed(
  id: string,
  disputed: boolean,
): Promise<DocumentNode> {
  return api
    .patch<DocumentNodeWire>(`/v1/documents/${encodeURIComponent(id)}`, {
      disputed,
    })
    .then(toNode)
    .then((node) => {
      scheduleAccountRulesMemoryRefresh();
      return node;
    });
}

/**
 * Overwrite a document's body (full-text, CAS-guarded). `baseline` is the version the edit
 * was based on; `null` writes unconditionally. A stale baseline returns
 * `{ ok: false, conflict: true }` with the live version.
 */
export function writeDocument(
  id: string,
  content: string,
  baseline: string | null,
): Promise<DocumentWriteResult> {
  return api
    .put<DocumentWriteWire>(`/v1/documents/${encodeURIComponent(id)}`, {
      content,
      baseline,
    })
    .then(toWriteResult)
    .then((result) => {
      if (result.ok && !result.conflict) scheduleAccountRulesMemoryRefresh();
      return result;
    });
}

/** Rename a document (content untouched). */
export function renameDocument(
  id: string,
  name: string,
): Promise<DocumentNode> {
  return api
    .patch<DocumentNodeWire>(`/v1/documents/${encodeURIComponent(id)}`, {
      name,
    })
    .then(toNode)
    .then((node) => {
      scheduleAccountRulesMemoryRefresh();
      return node;
    });
}

/** Soft-delete a document (and, for a folder, its subtree). */
export function deleteDocument(id: string): Promise<DocumentWriteResult> {
  return api
    .delete<DocumentWriteWire>(`/v1/documents/${encodeURIComponent(id)}`)
    .then(toWriteResult)
    .then((result) => {
      if (result.ok && !result.conflict) scheduleAccountRulesMemoryRefresh();
      return result;
    });
}

File: apps/server/agentcore/api/routes/documents.py
"""Documents API — tree CRUD; user rules expose ``always`` | ``on_demand`` apply_mode.

Self-only CRUD over the single ``documents`` content tree. ``apply_mode`` / ``description``
on the wire are **derived indexes** of the body's frontmatter (sole writable source).
Patching ``apply_mode`` edits frontmatter then re-derives — never a column-only write.
``frontmatter_error`` surfaces structural parse failure for the UI (unclosed fence).

Write-side always quota (闸在写侧): create / content edit / promote-to-always go through
the always-pool gate. Editing an existing always entry past the cap is allowed with
``quota_warning``; creating or promoting past the cap is refused (409).

``PATCH {disputed}`` is the 纠错通道 (「这条不对」): a user-only mark that stops an entry
from being injected / consulted while keeping it readable here. ``disputed_at`` rides
every node view so the UI can show and undo the mark.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from agentcore.api.dependencies import AuthUser, get_document_repo
from agentcore.db.models import Document
from agentcore.db.repositories import DocumentRepository
from agentcore.documents.description import maybe_schedule_description_fill
from agentcore.documents.frontmatter import (
    FrontmatterEditError,
    FrontmatterError,
    ParsedFrontmatter,
    frontmatter_error_message,
    parse_entry_frontmatter,
    set_entry_frontmatter,
)
from agentcore.documents.write_guards import is_ai_core_memory_leaf
from agentcore.memory import memory_version
from agentcore.memory.always_quota import (
    always_entry_chars,
    check_always_write,
    measure_always_usage,
)

router = APIRouter(prefix="/documents", tags=["documents"])

DocKind = Literal["folder", "document"]
DocRole = Literal["rule", "general"]
DocApplyMode = Literal["always", "on_demand"]


class DocumentNodeView(BaseModel):
    """A tree node's metadata (list rows — body omitted so a listing stays light)."""

    id: str
    parent_id: str | None
    folder_id: str | None
    kind: str
    role: str
    ai_maintained: bool
    apply_mode: str
    description: str
    name: str
    frontmatter_error: str | None = None
    # Chars counting toward the always pool (same meter as write-side gate); null if not always.
    always_chars: int | None = None
    # When the user marked this entry wrong (纠错通道). Set ⇒ the entry is kept and still
    # readable / editable here, but never injected and never offered to ``consult``.
    disputed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class DocumentDetailView(DocumentNodeView):
    """A node plus its markdown body and content-hash CAS tag (the editor's load payload)."""

    content: str
    version: str
    quota_warning: str | None = None


class DocumentCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=500)
    kind: DocKind = "document"
    role: DocRole = "general"
    content: str = ""
    # Default always for user rules (UI); written into frontmatter on create.
    apply_mode: DocApplyMode = "always"
    parent_id: str | None = None
    folder_id: str | None = None


class DocumentContentRequest(BaseModel):
    content: str
    baseline: str | None = None


class DocumentWriteResult(BaseModel):
    ok: bool
    version: str
    conflict: bool = False
    frontmatter_error: str | None = None
    quota_warning: str | None = None


class DocumentPatchRequest(BaseModel):
    """Rename, reparent, change apply (via frontmatter edit), and/or mark 这条不对."""

    name: str | None = Field(default=None, min_length=1, max_length=500)
    parent_id: str | None = None
    reparent: bool = False
    apply_mode: DocApplyMode | None = None
    # True = 「这条不对」(stop injecting, keep the entry); False = undo the mark.
    # Only ever set by an explicit user action in the memory UI.
    disputed: bool | None = None


class AlwaysQuotaView(BaseModel):
    """Always-pool usage for the UI meter (percentage + absolute chars)."""

    used_chars: int
    max_chars: int
    percent: float
    # ``used_chars == global_chars + project_chars`` (project context = global ∪ this project).
    global_chars: int
    project_chars: int


def _fm_error(doc: Document) -> str | None:
    if doc.kind != "document":
        return None
    return frontmatter_error_message(doc.content)


def _always_chars(doc: Document) -> int | None:
    """Pool chars for always-injected rule docs; null for everything else."""
    if doc.kind != "document" or doc.role != "rule" or doc.apply_mode != "always":
        return None
    return always_entry_chars(doc.content)


def _node(doc: Document) -> DocumentNodeView:
    return DocumentNodeView(
        id=doc.id,
        parent_id=doc.parent_id,
        folder_id=doc.folder_id,
        kind=doc.kind,
        role=doc.role,
        ai_maintained=doc.ai_maintained,
        apply_mode=doc.apply_mode,
        description=doc.description,
        name=doc.name,
        frontmatter_error=_fm_error(doc),
        always_chars=_always_chars(doc),
        disputed_at=doc.disputed_at,
        created_at=doc.created_at,
        updated_at=doc.updated_at,
    )


def _detail(doc: Document, *, quota_warning: str | None = None) -> DocumentDetailView:
    return DocumentDetailView(
        id=doc.id,
        parent_id=doc.parent_id,
        folder_id=doc.folder_id,
        kind=doc.kind,
        role=doc.role,
        ai_maintained=doc.ai_maintained,
        apply_mode=doc.apply_mode,
        description=doc.description,
        name=doc.name,
        frontmatter_error=_fm_error(doc),
        always_chars=_always_chars(doc),
        disputed_at=doc.disputed_at,
        content=doc.content,
        version=memory_version(doc.content),
        created_at=doc.created_at,
        updated_at=doc.updated_at,
        quota_warning=quota_warning,
    )


def _resolve_create_apply_mode(*, role: str, kind: str, apply_mode: DocApplyMode) -> str:
    """Only user-rule documents may be on_demand; everything else stays always."""
    if role == "rule" and kind == "document":
        return apply_mode
    return "always"


def _raise_quota_denied(message: str | None) -> None:
    raise HTTPException(
        status_code=409,
        detail={
            "code": "ALWAYS_QUOTA_EXCEEDED",
            "message": message or "常驻条目配额已满",
        },
    )


async def _preview_create_body(
    *, role: str, kind: str, apply_mode: DocApplyMode, content: str
) -> tuple[str, bool]:
    """Return (body_as_stored, is_always_rule) for quota projection before create."""
    if kind != "document" or role != "rule":
        return content, False
    mode = _resolve_create_apply_mode(role=role, kind=kind, apply_mode=apply_mode)
    try:
        body = set_entry_frontmatter(content, apply=mode)  # type: ignore[arg-type]
    except FrontmatterEditError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return body, mode == "always"


@router.get("/always-quota", response_model=AlwaysQuotaView)
async def get_always_quota(
    user: AuthUser,
    folder_id: str | None = None,
    repo: DocumentRepository = Depends(get_document_repo),
) -> AlwaysQuotaView:
    """Always-pool usage for the injection context (global + optional project)."""
    usage = await measure_always_usage(repo, user.user_id, folder_id=folder_id)
    return AlwaysQuotaView(
        used_chars=usage.used_chars,
        max_chars=usage.max_chars,
        percent=usage.percent,
        global_chars=usage.global_chars,
        project_chars=usage.project_chars,
    )


@router.get("", response_model=list[DocumentNodeView])
async def list_documents(
    user: AuthUser,
    parent_id: str | None = None,
    repo: DocumentRepository = Depends(get_document_repo),
) -> list[DocumentNodeView]:
    """List a folder's direct children (``parent_id`` omitted = the user's top-level nodes)."""
    nodes = await repo.list_children(user.user_id, parent_id=parent_id)
    return [_node(n) for n in nodes]


@router.post("", response_model=DocumentDetailView)
async def create_document(
    body: DocumentCreateRequest,
    user: AuthUser,
    repo: DocumentRepository = Depends(get_document_repo),
) -> DocumentDetailView:
    """Create a tree node (always ``ai_maintained=false`` — user-owned).

    A child inherits its parent's ``folder_id`` scope; a root node takes the requested scope.
    New ``role='rule'`` documents with no parent land under ``AgentCore/规则/`` (§5.0).
    """
    folder_id = body.folder_id
    parent_id = body.parent_id
    if parent_id is not None:
        parent = await repo.get(parent_id, user_id=user.user_id)
        if parent is None:
            raise HTTPException(status_code=404, detail="parent not found")
        if parent.kind != "folder":
            raise HTTPException(status_code=400, detail="parent is not a folder")
        folder_id = parent.folder_id
    elif body.role == "rule" and body.kind == "document":
        rules_dir = await repo.ensure_rules_dir(user.user_id, folder_id)
        parent_id = rules_dir.id
        folder_id = rules_dir.folder_id

    preview_body, is_always = await _preview_create_body(
        role=body.role, kind=body.kind, apply_mode=body.apply_mode, content=body.content
    )
    if is_always:
        decision = await check_always_write(
            repo,
            user.user_id,
            folder_id=folder_id,
            writer="user",
            editing_existing_always=False,
            exclude_id=None,
            new_content=preview_body,
            new_is_always=True,
        )
        if not decision.allowed:
            _raise_quota_denied(decision.message)

    try:
        doc = await repo.create(
            user.user_id,
            name=body.name,
            parent_id=parent_id,
            folder_id=folder_id,
            kind=body.kind,
            role=body.role,
            ai_maintained=False,
            apply_mode=_resolve_create_apply_mode(
                role=body.role, kind=body.kind, apply_mode=body.apply_mode
            ),
            content=body.content if body.kind == "document" else "",
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    # Async empty-description fill — never blocks / fails the save (定案).
    maybe_schedule_description_fill(
        document_id=doc.id,
        user_id=user.user_id,
        kind=doc.kind,
        description=doc.description or "",
        content=doc.content or "",
    )
    return _detail(doc)


@router.get("/{document_id}", response_model=DocumentDetailView)
async def get_document(
    document_id: str,
    user: AuthUser,
    repo: DocumentRepository = Depends(get_document_repo),
) -> DocumentDetailView:
    doc = await repo.get(document_id, user_id=user.user_id)
    if doc is None:
        raise HTTPException(status_code=404, detail="document not found")
    return _detail(doc)


@router.put("/{document_id}", response_model=DocumentWriteResult)
async def update_document_content(
    document_id: str,
    body: DocumentContentRequest,
    user: AuthUser,
    repo: DocumentRepository = Depends(get_document_repo),
) -> DocumentWriteResult:
    """Overwrite a document's body (CAS-guarded; conflict instead of clobber)."""
    doc = await repo.get(document_id, user_id=user.user_id)
    if doc is None:
        raise HTTPException(status_code=404, detail="document not found")
    current_version = memory_version(doc.content)
    if body.baseline is not None and body.baseline != current_version:
        return DocumentWriteResult(
            ok=False,
            version=current_version,
            conflict=True,
            frontmatter_error=_fm_error(doc),
        )

    quota_warning: str | None = None
    if doc.kind == "document" and doc.role == "rule":
        parsed = parse_entry_frontmatter(body.content)
        if isinstance(parsed, FrontmatterError):
            new_is_always = False
        elif isinstance(parsed, ParsedFrontmatter):
            new_is_always = parsed.apply == "always"
        else:
            new_is_always = False

        editing_existing_always = doc.apply_mode == "always"
        decision = await check_always_write(
            repo,
            user.user_id,
            folder_id=doc.folder_id,
            writer="user",
            editing_existing_always=editing_existing_always,
            exclude_id=doc.id,
            new_content=body.content,
            new_is_always=new_is_always,
        )
        if not decision.allowed:
            _raise_quota_denied(decision.message)
        quota_warning = decision.warning

    updated = await repo.update_content(document_id, user_id=user.user_id, content=body.content)
    assert updated is not None
    # Clear → regenerate; non-empty description never auto-overwritten (定案).
    maybe_schedule_description_fill(
        document_id=updated.id,
        user_id=user.user_id,
        kind=updated.kind,
        description=updated.description or "",
        content=updated.content or "",
    )
    return DocumentWriteResult(
        ok=True,
        version=memory_version(body.content),
        frontmatter_error=_fm_error(updated),
        quota_warning=quota_warning,
    )


@router.patch("/{document_id}", response_model=DocumentDetailView)
async def patch_document(
    document_id: str,
    body: DocumentPatchRequest,
    user: AuthUser,
    repo: DocumentRepository = Depends(get_document_repo),
) -> DocumentDetailView:
    """Rename, reparent, set apply via frontmatter, and/or mark the entry wrong.

    Set ``reparent`` to apply ``parent_id``. ``disputed`` is the 纠错通道: the user says
    「这条不对」in the memory UI and the entry stops being injected / consulted while
    staying on disk (no silent delete, and the mark survives later AI rewrites because it
    lives in a column, not in the body). Nothing infers this from conversation text.
    """
    doc = await repo.get(document_id, user_id=user.user_id)
    if doc is None:
        raise HTTPException(status_code=404, detail="document not found")
    if body.name is not None:
        await repo.rename(document_id, user_id=user.user_id, name=body.name)
    if body.reparent:
        new_folder = doc.folder_id
        if body.parent_id is not None:
            parent = await repo.get(body.parent_id, user_id=user.user_id)
            if parent is None:
                raise HTTPException(status_code=404, detail="parent not found")
            if parent.kind != "folder":
                raise HTTPException(status_code=400, detail="parent is not a folder")
            new_folder = parent.folder_id
        await repo.move(
            document_id, user_id=user.user_id, parent_id=body.parent_id, folder_id=new_folder
        )
    quota_warning: str | None = None
    if body.apply_mode is not None:
        current = await repo.get(document_id, user_id=user.user_id)
        assert current is not None
        if current.ai_maintained:
            raise HTTPException(
                status_code=400, detail="cannot change apply_mode of AI-maintained documents"
            )
        if current.role != "rule" or current.kind != "document":
            raise HTTPException(
                status_code=400, detail="apply_mode only applies to rule documents"
            )
        if body.apply_mode == "always" and current.apply_mode != "always":
            try:
                preview = set_entry_frontmatter(current.content, apply="always")
            except FrontmatterEditError as exc:
                raise HTTPException(status_code=400, detail=str(exc)) from exc
            decision = await check_always_write(
                repo,
                user.user_id,
                folder_id=current.folder_id,
                writer="user",
                editing_existing_always=False,
                exclude_id=current.id,
                new_content=preview,
                new_is_always=True,
            )
            if not decision.allowed:
                _raise_quota_denied(decision.message)
            quota_warning = decision.warning
        try:
            await repo.update_apply_mode(
                document_id, user_id=user.user_id, apply_mode=body.apply_mode
            )
        except FrontmatterEditError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
    if body.disputed is not None:
        marked = await repo.set_disputed(
            document_id, user_id=user.user_id, disputed=body.disputed
        )
        if marked is None:
            raise HTTPException(status_code=400, detail="cannot dispute this node")
    refreshed = await repo.get(document_id, user_id=user.user_id)
    assert refreshed is not None
    return _detail(refreshed, quota_warning=quota_warning)


@router.delete("/{document_id}", response_model=DocumentWriteResult)
async def delete_document(
    document_id: str,
    user: AuthUser,
    repo: DocumentRepository = Depends(get_document_repo),
) -> DocumentWriteResult:
    """Soft-delete a node and (for a folder) its whole subtree.

    AI-maintained core leaves (偏好 / 画像 / 导航) keep their protocol names, so
    this DELETE is refused. Empty the body instead (``PUT …/memory/files/{kind}``
    with empty content) — injection skips the empty note; the list still shows
    a placeholder. On-demand AI topics and user-owned entries remain deletable.
    """
    doc = await repo.get(document_id, user_id=user.user_id)
    if doc is None:
        raise HTTPException(status_code=404, detail="document not found")
    if is_ai_core_memory_leaf(name=doc.name, ai_maintained=doc.ai_maintained):
        raise HTTPException(
            status_code=400,
            detail="cannot delete AI-maintained core memory leaves",
        )
    ok = await repo.soft_delete(document_id, user_id=user.user_id)
    if not ok:
        raise HTTPException(status_code=404, detail="document not found")
    return DocumentWriteResult(ok=True, version=memory_version(""))
